File: src/handlers/lib/refresh.rs
use crate::{
    config::Config,
    crypto,
    crypto::jwt::JWT,
    handlers::{lib::token::LoginResponse, Error},
    hook::{HookEvent, Webhook},
    models::{refresh_token::get_refresh_token_by_token, user},
    operator_signature::OperatorSignature,
};
use diesel::{
    pg::PgConnection,
    r2d2::{ConnectionManager, PooledConnection},
};
use log::error;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, GraphQLInputObject)]
pub struct RefreshForm {
    #[graphql(name = "refresh_token")]
    pub refresh_token: String,
}

pub fn refresh(config: &Config, connection: &PooledConnection<ConnectionManager<PgConnection>>, operator_signature: OperatorSignature, refresh_form: RefreshForm) -> Result<LoginResponse, Error> {
    let internal_error = Error::new(500, json!({"code": "internal_server_error"}), "Internal Server Error".to_string());

    let invalid_refresh_token = Error::new(400, json!({"code": "invalid_refresh_token"}), "Invalid Refresh Token".to_string());

    let refresh_token = get_refresh_token_by_token(refresh_form.refresh_token, &connection);

    if refresh_token.is_err() {
        let err = refresh_token.err().unwrap();

        error!("{:?}", err);

        return Err(invalid_refresh_token);
    }

    let mut refresh_token = refresh_token.unwrap();

    let token = crypto::secure_token(50);

    refresh_token.token = token;

    let refresh_token = refresh_token.save(&connection);

    if refresh_token.is_err() {
        let err = refresh_token.err().unwrap();

        error!("{:?}", err);

        return Err(internal_error);
    }

    let refresh_token = refresh_token.unwrap();

    let user = user::get_by_id(refresh_token.user_id, &connection);

    if user.is_err() {
        let err = user.err().unwrap();

        error!("{:?}", err);

        return Err(internal_error);
    }

    let user = user.unwrap();

    let payload = json!({
        "event": HookEvent::Login,
        "provider": "email",
        "user": user,
    });

    let hook = Webhook::new(HookEvent::Login, payload, config.clone(), operator_signature);

    let hook_response = hook.trigger();

    if hook_response.is_err() {
        return Err(Error::from(hook_response.err().unwrap()));
    }

    let hook_response = hook_response.unwrap();

    let jwt = JWT::new(&user, config.aud.clone(), hook_response);

    let jwt = jwt.sign(config);

    if jwt.is_err() {
        let err = jwt.err().unwrap();

        error!("{:?}", err);

        return Err(internal_error);
    }

    let jwt = jwt.unwrap();

    return Ok(LoginResponse {
        access_token: jwt,
        refresh_token: refresh_token.token,
    });
}

File: handlers/lib/refresh.go
package lib

import (
	"github.com/sirupsen/logrus"
	"github.com/thanhpk/randstr"
	"github.com/zolamk/trust/config"
	"github.com/zolamk/trust/errors"
	"github.com/zolamk/trust/hook"
	"github.com/zolamk/trust/jwt"
	"github.com/zolamk/trust/model"
	"gorm.io/gorm"
)

func RefreshToken(db *gorm.DB, config *config.Config, rt string) (*model.LoginResponse, error) {

	var refresh_token model.RefreshToken

	if tx := db.Joins("User").First(&refresh_token, "token = ?", rt); tx.Error != nil {
		if tx.Error == gorm.ErrRecordNotFound {
			return nil, errors.RefreshTokenNotFound
		}
		return nil, errors.Internal
	}

	user := refresh_token.User

	refresh_token.Token = randstr.String(50)

	payload := &map[string]interface{}{
		"event":    "login",
		"provider": "email",
		"user":     user,
	}

	hook_response, err := hook.TriggerHook("login", payload, config)

	if err != nil {
		logrus.Error(err)
		return nil, errors.WebHook
	}

	token := jwt.New(user, hook_response, config.JWT)

	signed_token, err := token.Sign()

	if err != nil {
		logrus.Error(err)
		return nil, errors.Internal
	}

	if err := refresh_token.Save(db); err != nil {
		logrus.Error(err)
		return nil, errors.Internal
	}

	return &model.LoginResponse{
		AccessToken:  signed_token,
		RefreshToken: refresh_token.Token,
		ID:           user.ID,
	}, nil
}
